Swift: Add variable-capture flow

MathiasVP opened the pull request:
Commit-by-commit review recommended. I've kept most of Nora's commits, but note that many of the things added in those commits are actually changed by my own commits afterwards 😂.

DCA looks fine 🎉.

The QLDoc check fails because the shared consistency queries don't have any QLDoc. This will only fail this one time because this PR is introducing those consistency queries.

aschackmull commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -396,6 +419,13 @@ private module ParameterNodes {
     override ParamDecl getParameter() { result = param }
   }
 
+  class ClosureSelfParameterNode extends ParameterNodeImpl, ClosureSelfReferenceNode {
```
This looks wrong, but I might be misinterpreting some detail of the Swift AST. I'd guess that a `ClosureExpr` corresponds roughly to a `LambdaExpr` in Java, which means that this node belongs to the surrounding scope of the closure and not the scope of the closure, which contradicts it being a parameter of the closure.

MathiasVP replied:
Oops you're right. This doesn't make sense at all 👀

MathiasVP replied:
Oh yes this doesn't make any sense!

MathiasVP replied:
Fixed in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb.

aschackmull commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -785,6 +827,156 @@ private module OutNodes {
+
+  BasicBlock getABasicBlockSuccessor(BasicBlock bb) { result = bb.(B::BasicBlock).getASuccessor() }
+
+  //TODO: support capture of `this` in lambdas
```
If I happened to have got the semantics of a `this` reference in a swift lambda wrong in my comments above, then this todo is irrelevant. But if I got it right, then the todo is fine and indeed relevant.

aschackmull replied:
Unless of course the `this` variable in Swift is a `VarDecl` in which case you might already be handling this-capture?

MathiasVP replied:
Oh, I had missed this TODO comment. I believe `this` is a `VarDecl` in Swift, but let me double check that real quick 👀

MathiasVP replied:
Indeed, a `self` (Swift is one of those cool Smalltalk-like languages 😎) is a `VarDecl`, and a `CapturedDecl` exists for `self` when it's captured in a closure. So this TODO is irrelevant.

aschackmull replied:
Do you have a test case for a captured `this`? Something like https://github.com/github/codeql/blob/53d92d58fc27f37888862593dfb6adbd3986aae6/java/ql/test/library-tests/dataflow/capture/B.java#L20-L36 (those missing results in that Java test is what it's about).

MathiasVP replied:
I've removed the TODO in 78e08cf63ccadde6cdb7f478421460bc816dcf7a. I don't think there's such a test, but I'll make sure to add such a test. There might be something else that's ruining this flow 😂.

MathiasVP replied:
I've added a version of the above test in 784bb72b3340070a3525a18e7947dfeb3562598b. It turns out we don't properly capture `self`, but this is because of an extractor issue (I've created an internal issue for this now).

aschackmull commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -785,6 +827,156 @@ private module OutNodes {
+  or
+  result.(CaptureFlow::ThisParameterNode).getCallable() = n.(ClosureSelfParameterNode).getClosure()
+  or
+  result.(CaptureFlow::MallocNode).getClosureExpr() = n.getCfgNode().getNode().asAstNode() // TODO: figure out why the java version had PostUpdateNode logic here
```
I believe I answered this todo on the other PR.

MathiasVP replied:
Yeah, I basically didn't touch this since I didn't bother investigating captures through constructors in this PR. But I can do something better than keep the TODO here, for sure.

MathiasVP replied:
From some local experimentation with the Swift compiler I gather that constructors cannot perform variable capturing in Swift (by design). So I've removed this code in b41ec37993f46f24fafef409dc07a3db194c075d.

aschackmull commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPublic.qll:
```diff
@@ -51,6 +51,11 @@ class Node extends TNode {
+  /**
+   * Gets the parameter that corresponds to this node, if any.
+   */
+  ParamDecl getParameter() { none() }
```
Shouldn't this predicate be called `asParameter` for consistency? I.e. to indicate that it corresponds to a cast more than a getter.
```suggestion
  ParamDecl asParameter() { result = this.(ParameterNode).getParameter() }
```

MathiasVP replied:
Indeed, it should. I'll fix that up in a commit since we need to rename a use of it as well.

MathiasVP replied:
Turns out there was a lot of changes necessary here. There seems to have been some confusion about `asParameter` (for getting the node corresponding to a parameter) and `getParameter` (for any kind of dataflow node that happened to be defined in terms of a parameter). This has been fixed in 951b6beeb150f5b0db1e20a50ce52a2094d76894. Thanks for bringing this to my attention!

MathiasVP commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -785,6 +827,156 @@ private module OutNodes {
+    def.assigns(any(CfgNode cfg | cfg.getNode().asAstNode() = e1))
+  )
+  or
+  e2.(Pattern).getImmediateMatchingExpr() = e1
```
Note that `e2` is of type `CaptureInput::Expr`, and `CaptureInput::Expr` is actually an [AstNode](https://github.com/github/codeql/pull/14577/files/862de152a1308d02ee43fa01f3cf764357303f7f#diff-514dd0988f7b47e40ff8c1e66ea886d1f947d047da108270ed0ddec75c2856c1R889).

aschackmull replied:
Yes, that's why compilation doesn't fail. I'm asking whether `Expr` and `Pattern` might be sufficiently disjoint (even though both are subtypes of `AstNode`) that this will block flow?

MathiasVP replied:
I think I'll leave this to another PR. I've discovered that the `WriteDefinition::assigns` predicate doesn't really do the right thing at all. For a variable declaration such as:
```
var x = 0
```
the `WriteDefinition::assigns(value)` predicate holds where `value` is the left-hand side of the declaration (i.e., the lvalue of `x`) instead of the right-hand side (i.e., `0`). Once that's fixed (which I'll do in a later PR) we can remove this disjunct and have `CaptureInput::Expr` be `Expr`.

geoffw0 reviewed (commented):
Thank you for your review @aschackmull .  I've added a few comments but I think you have more insight here than I do.

I'm pleased to see these changes fix multiple `[NOT DETECTED]` and `MISSING` results from existing tests, which suggests things are working well enough to make a difference.

> I didn't bother investigating captures through constructors in this PR

Is there an issue for / test showing this?

MathiasVP commented:
> > I didn't bother investigating captures through constructors in this PR
> 
> Is there an issue for / test showing this?

For sure not an issue. I haven't checked if we have a test for it (I suspect not, but I'll add one before this PR is merged).

geoffw0 commented:
Forgot to add, DCA LGTM as well.

MathiasVP commented on swift/ql/test/library-tests/dataflow/dataflow/test.swift:
```diff
@@ -129,7 +129,7 @@ func forwarder() {
         return 0
     })
-    sink(arg: clean)
+    sink(arg: clean) // $ SPURIOUS: flow=128
```
Yes. We're missing call contexts for the call above:
```cpp
forward(arg: source(), lambda: {
  (i: Int) -> Int in
    return 0
})
```
so we actually get flow like:
```
source() (i.e., the argument) -> arg (i.e., the parameter of `forward`) -> arg (i.e., the argument of lambda(arg)) -> i (the parameter of on line 123) -> i (the return on line 124) -> clean
```

There can be two explanations of this:
- Swift doesn't implement dispatch through call contexts (see `mayBenefitFromCallContext` and `viableImplInCallContext` in `DataFlowDispatch`).
- The call contexts exists, but is being thrown away because there's an issue with an enclosing callable. I actually believe this is the case based on @aschackmull's feedback [here](https://github.com/github/codeql/pull/14577#discussion_r1371503179).

I'll look into this as well.

MathiasVP replied:
This was fixed in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb (the answer was indeed a version of case 2).

geoffw0 replied:
We've lost two good results in that change - and they look like fairly straightforward ones, unless I'm missing something?

MathiasVP replied:
We're losing this one:
```swift
func setEscape() {
  let x = source("setEscape", 0)
  escape = {
    sink(x) // $ MISSING: hasValueFlow=setEscape
    return x + 1
  }
}
```
because we don't model global flow in Swift. So we don't manage to flow from the definition of `escape` to the place where it's actually called:
```swift
func callEscape() {
  setEscape()
  sink(escape?()) // $ MISSING: hasTaintFlow=setEscape
}
```
In an earlier commit we accidentially got this flow because we flowed from `let x = source("setEscape", 0)` and straight into the closure (which violates all kinds of dataflow invariants), but when I fixed that in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb the result disappears again, and we're only gonna get it back if we implement global flow in Swift.

For this one:
```swift
func captureList() {
  let y: Int = source("captureList", 123);
  { [x = hello()] () in
     sink(x) // $ MISSING: hasValueFlow=hello
     sink(y) // $ MISSING: hasValueFlow=captureList
  }()
}
```
I feel like we should be able to catch `sink(y)` (but for some reason we still don't). We don't catch the `sink(x)` one because we're missing an SSA definition for `x = hello()` and that work is blocked on https://github.com/github/codeql/pull/14567 (which I've promised Paolo to look at after this is merged).

MathiasVP replied:
Yep I'm comfortable calling the the above an issue with capture lists. I've added a version of this test in 65e13aa5edf5a8ff3446e41c0aab3f4ed2ce3266 without a capture list, and that works fine. So I think we should delay further work on this until https://github.com/github/codeql/pull/14567 is merged.

aschackmull commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -785,6 +827,155 @@ private module OutNodes {
+        variable = np.getVarDecl() and
+        source = np.getMatchingExpr()
+      )
+      // TODO: support multiple variables in LHS of =, in both of above cases.
```
And a test, if you don't have one.

MathiasVP replied:
Indeed. I don't think there's a test for it. I'll add one.

First I need to understand what the TODO is actually saying. If `multiple variables in LHS of =` is referring to a destructuring such as `var (a, b) = myPair` then I wouldn't expect this to be a `VariableWrite`. Instead, I'd expect this to be a read step from `myPair` to `a` (and similarl for `b`) just like we do for normal dataflow.

geoffw0 replied:
> `var (a, b) = myPair`

That was my understanding of what it's saying as well.

MathiasVP replied:
I've deleted the TODO in 63525a9d9efb0a1683dd7d10c3eb455e5b6c2231. There's some potential for cleanup, and I've created an internal issue for this now.

MathiasVP commented on swift/ql/lib/codeql/swift/dataflow/internal/DataFlowDispatch.qll:
```diff
@@ -293,12 +293,14 @@ private module Cached {
     TThisParameter() or
-    TPositionalParameter(int n) { n = any(Argument arg).getIndex() }
+    TPositionalParameter(int n) { n = any(Argument arg).getIndex() } or
+    TClosureSelfParameter()
```
In 11194e574ce9241126cac1e09cb7310583cc16cd we're now re-using the this parameter (and argument). Is that what you had in mind?

This seems to add a bunch of spurious call-context related FPs. But that may be related to https://github.com/github/codeql/pull/14577#discussion_r1371503179.

aschackmull replied:
Pretty much, yes. But it's hard to judge entirely since `ClosureSelfParameterNode` is still very wrong.

MathiasVP replied:
Yep. I'm about to push a fix for that now 🤞

MathiasVP replied:
Fixed the `ClosureSelfParameterNode` in 2465cc20f0108a9b80c85e438695cb1f43b2f3eb.

MathiasVP commented on swift/ql/test/TestUtilities/InlineFlowTest.qll:
```diff
@@ -0,0 +1,117 @@
+ *     let s = source()
+ *     sink(s) // $ hasValueFlow
+ *     let t = "foo" + taint()
+ *     sink(t); // $ hasTaintFlow
```
Yeah, this is really a much better way to use inline expectations tests. I think that's how other languages mark dataflow tests as well.

geoffw0 replied:
Yeah, I'll write up an issue for doing this (for Swift, that is).

MathiasVP commented on swift/ql/test/library-tests/dataflow/capture/closures.swift:
```diff
@@ -0,0 +1,160 @@
+  sink(getX()) // $ MISSING: hasTaintFlow=sharedCapture
+  doubleIncrX()
+  return getX()
+}
```
Fixed in 2ad121a8a571e2f926ffb782ff6997c9f999101c.

MathiasVP commented on swift/ql/test/library-tests/dataflow/capture/closures.swift:
```diff
@@ -0,0 +1,160 @@
+  let r = { x }
+  g!(1)
+  return r
+}
```
I did option 4 in 96a37f3a3c34eca44c4a36bfeff762e1770b8086.

MathiasVP commented:
> For sure not an issue. I haven't checked if we have a test for it (I suspect not, but I'll add one before this PR is merged).

I've now investigated this. It seems like this is simply irrelevant for Swift because of fantastic language design choices 😂 (see https://github.com/github/codeql/pull/14577/commits/b41ec37993f46f24fafef409dc07a3db194c075d for more details).

MathiasVP commented:
@geoffw0 @aschackmull I believe I've addressed all review comments. There are a few unresolved ones (https://github.com/github/codeql/pull/14577#discussion_r1371511680 and https://github.com/github/codeql/pull/14577#discussion_r1371518127) that are blocked on work that I don't want to include in this PR (but which we now have issues/other ongoing PRs for).

aschackmull commented:
SGTM. I'll leave approval to @geoffw0 as I haven't looked at the tests at all.

geoffw0 reviewed (approved):
Yes I'm happy.  There's quite a bit of follow-up work but I agree it's time to get this merged and start on incremental improvements.

geoffw0 commented:
:tada:
